098-using-image-volumes-to-improve-extensibility-of-Strimzi-operands.md

```diff
@@ -0,0 +1,320 @@
+The advantage of the tag is that you can always use the latest version of the plugin without updating the custom resource.
+But it will not be deterministic what is the exact version used.
+
+The following example shows the new YAML of the `KafkaConnect` resource:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaConnect
+metadata:
+  # ...
+spec:
+  # ...
+  plugins:
+    - name: echo-sink-connector
+      artifacts:
+        - type: image
+          reference: ghcr.io/scholzj/echo-sink@sha256:7078b3ccbc0d6e76fefb832dd6e4bb6b704f83611428c9d5b060ab0b6c3e8712
+          pullPolicy: Always
+```
+
+Based on the `KafkaConnect` custom resource, the Strimzi Cluster operator will add the artifacts as image volumes to the Pod definitions.
+The volumes will be named using the following schema: `plugin-<pluginName>-<imageReferenceHashstub>`.
+This will ensure the volume names are unique.
+For example:
+
+```yaml
+  # ...
+  volumes:
+    # ...
+    - image:
+        pullPolicy: Always
+        reference: ghcr.io/scholzj/echo-sink@sha256:7078b3ccbc0d6e76fefb832dd6e4bb6b704f83611428c9d5b060ab0b6c3e8712
+  # ...
+```
+
+This volume will be afterwards mounted to the Kafka Connect container into a subdirectory structure on the `/opt/kafka/plugins/` path which is used when adding connectors to the custom container image as well.
+The one or more containers will be mounted to `/opt/kafka/plugins/<pluginName>/<imageReferenceHashstub>/`.
+For example:
+
+```yaml
+    # ...
+    volumeMounts:
+      - mountPath: /opt/kafka/plugins/echo-sink-connector/9e1cec06
+        name: plugin-echo-sink-connector-9e1cec06
+    # ...
+```
+
+When the same plugin name is used in `.spec.build.plugins`, the artifacts from both sections will end up in the same directory tree `/opt/kafka/plugins/<pluginName>/`.
+
+### Other plugins
+
+Mounting other plugins in other components is not as common as Kafka Connect connector plugins.
+Most of the other plugins also need to be added to the regular class path of the application rather than to a special `plugin.path`.
+Therefore mounting other plugins does not use a special API.
+It uses the existing additional volumes instead and adds a new volume type named `image` to it to allow mounting image volumes.
+This allows the users to mount the container image into an directory inside the `/mnt` path.
+
+Afterwards, they can use an environment variable to add the JARs from the mounted container image to the class path.
+For Kafka-based operands and Cruise Control the environment variable to use is `CLASSPATH`.
+For Strimzi operands (User and Topic operators), the environment variable to use is `JAVA_CLASSPATH`.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  # ...
+spec:
+  kafka:
+    # ...
+    template:
+      kafkaContainer:
+        volumeMounts:
+          - name: my-authorizer
+            mountPath: /mnt/my-authorizer/
```

Comment:
Should the proposal document a recommended mount point that users should use?  I suggest this to avoid conflicts between user configuration and future Strimzi enhancement (that might use mount points of their own).

Reply:
The _additional volumes_ (an existing functionality to which I just add a new volume type) allow mounting only into the `/mnt` path. So that should take care of it. I tried to make it clear in the proposal.